R2: REMS-registration

samuli opened the pull request:
Käyttäjän rekisteröinti REMS-luvituspalveluun.

Uutta:
~~- ChoiceAuth pluginiin uusi metodi getPostLoginLightboxRoute, jonka kautta avataan rekisteröintilomake Suomifi-tunnistautumisen jälkeen.~~
- Rekisteröintilomake uusia ja palaavia käyttäjiä varten (riippuen siitä onko käyttäjä jo rekisteröity REMSiin).

TODO
- [x] fi-tekstit
- [x] ~~Käännökset~~ lisätään myöhemmin


**Riippuvuudet**
- [x] https://github.com/NatLibFi/NDL-VuFind2/pull/1694
- [x] https://github.com/NatLibFi/NDL-VuFind2/pull/1700
- [x] https://github.com/vufind-org/vufind/pull/1724

Commit:
[ALLI-5805] REMS registration.

Commit:
Add post-login lightbox handling to templates.

Commit:
REMS registration form styling.

Commit:
Feedback form styling.

Commit:
Remove unused method. Use R2Service. Update RemsService namespace.

Commit:
Remove unused.

Commit:
Update FeedbackForms.yaml.sample

Commit:
Revert change.

Commit:
Remove style.

Commit:
Add translations.

Commit:
phpcsfixer.

EreMaijala commented on local/config/finna/FeedbackForms.yaml.sample:
```diff
@@ -51,6 +51,8 @@
 #   or a different set of fields is configured in 'fields' list):
```
Olisi siistimpi, jos R2:n lomakkeet määriteltäisiin irrallaan FeedbackForms-lomakkeista. Ja toiminto voisi olla erillinen ja vain periä normaalin feedbackin. Silloin ei tarvitsisi tavallisen feedbackin puolelle sekoittaa R2:n maagisia tunnisteita yms.

samuli replied:
Erottelin R2-lomakkeiden käsittelyn R2FeedbackControlleriin, joka käsittelee R2Form:eja (joilla oma asetustiedostostosto). Tämä vaatii pienen muutoksen yläjuoksulle, mihin voin laittaa PR:n jos mennään tällä ratkaisulla:
https://github.com/NatLibFi/NDL-VuFind2/pull/1698/files#diff-ca0eb2be2fe3bb9c784e596a8538a127R30

EreMaijala replied:
Joo, parempi. Yläjuoksulle noin, tai sitten jopa erillinen metodi, joka luo formin.

samuli replied:
Lisäsin yläjuoksun PR:n riippuvuuksiin.

samuli replied:
Tällaisen jouduin vielä lisäämään Finnan FeedbackControlleriin koska formin action määritellään VuFindin templatessa. 
https://github.com/NatLibFi/NDL-VuFind2/pull/1698/files#diff-a038c3a11e015d3817cc49cf30ed702cR78
Tämän voisi välttää muuttamalla yläjuoksun controlleria niin että action asetetaan jo siellä.

EreMaijala commented on local/languages/finna/fi.ini:
```diff
@@ -682,6 +682,25 @@ Public List Visibility = "Näkyvyys"
 Published In Issue = "Julkaisussa"
 Publisher or Distributor Number = "Julkaisijan tunnus"
+R2_register_error = "Rekisteröityminen Finnan tutkijasaliin epäonnistui."
+R2_register_form_datasource = "(tiedot saatu suomi.fi-tunnistuspalvelusta)"
```
Suomi.fi kirjoitetaan ainakin suomi.fi:ssä isolla alkukirjaimella.

EreMaijala commented on local/languages/finna/fi.ini:
```diff
@@ -682,6 +682,25 @@ Public List Visibility = "Näkyvyys"
+R2_register_form_datasource = "(tiedot saatu suomi.fi-tunnistuspalvelusta)"
+R2_register_form_title = "Käyttösitoumus Finnan tutkijasalin sisältöihin"
+R2_register_form_help_post_html = "Antamasi tiedot rekisteröidään Finnan käyttäjärekisteriin ja tutkijasalin luvituspalveluun.<br>Lue lisää <a href="%%url%%" data-lightbox-ignore>tietosuojaselosteesta</a>."
+R2_register_form_help_pre_html = "<p><a href="%%url%%" data-lightbox-ignore>Lue lisää Finna-tutkijasalista</a></p><h4>Finnan tutkijasali aineistotiedon käyttöehdot</h4><p><strong>Finnan tutkijasalin  sisältämiä aineistotietoja (kuvailutietoa, luettelointitietoa, metatietoja) voi käyttää tietosuoja-asetuksen ja tietosuojalain asettamin rajoituksin.</strong> Käyttäjällä on oikeus hakea ja selata ai ...
```
Puuttuuko tästä jotain tms.: "Finnan tutkijasali aineistotiedon käyttöehdot"? Kuulostaa oudolta.

Tuplaväli pois: "tutkijasalin  sisältämiä"

samuli replied:
Muutin "Finna Tutkijasalin aineistotiedon...".

EreMaijala commented on local/languages/finna/fi.ini:
```diff
@@ -682,6 +682,25 @@ Public List Visibility = "Näkyvyys"
 Publisher or Distributor Number = "Julkaisijan tunnus"
+R2_register_error = "Rekisteröityminen Finnan tutkijasaliin epäonnistui."
+R2_register_form_datasource = "(tiedot saatu suomi.fi-tunnistuspalvelusta)"
+R2_register_form_title = "Käyttösitoumus Finnan tutkijasalin sisältöihin"
```
Kirjoitusasu joko "Finnan tutkijasali" tai "Finna-tutkijasali", nyt on molempia.

samuli replied:
Muutettu muotoon "Finna Tutkijasali".

EreMaijala commented on local/languages/finna/fi.ini:
```diff
@@ -682,6 +682,25 @@ Public List Visibility = "Näkyvyys"
+R2_register_form_license_accept = "Sitoudun metatiedon ja aineiston käyttöehtoihin"
+R2_register_form_usage = "Käyttötarkoituksesi aineistotiedolle"
+R2_register_form_usage_help_html = "Käyttäjä sitoutuu käyttämään aineistoja ilmoittamansa käyttötarkoituksen mukaisesti tällä kirjautumiskerralla.<span class="tooltip-r2-usage-purpose" data-toggle="tooltip" data-placement="auto" data-html="true" data-original-title="%%title%%"><i class="fa fa-info-big"></i></span>"
+R2_register_form_usage_help_tooltip_html = "<h4>Käyttötarkoitus</h4><p>Valitse parhaiten käyttötarkoitustasi kuvaava yllä olevasta listasta.</p><p>Käyttötarkoitus tulee ilmoittaa lain mukaan aina tietosuojatun aineiston käytön yhteydessä. Myös  aineistotietojen käyttö edellyttää käyttötarkoituksen ilmoittamista jokaisella käyttökerralla.</p><p>Käyttötarkoitus on selvitettävä riippumatta tietojen  ...
```
Tuplaväli: "Myös  aineistotietojen"

EreMaijala commented on module/Finna/src/Finna/Controller/R2ControllerTrait.php:
```diff
@@ -0,0 +1,202 @@
+            }
+
+            // Take firstname and lastname from profile if available
+            $firstname = !empty($user->firstname)
```
Jos nämä kentät ovat read only, ei niitä pidä ottaa parametreista koskaan, vai?

samuli replied:
Jeps, muutin ottamaan aina profiilista.

EreMaijala commented on module/Finna/src/Finna/Controller/R2ControllerTrait.php:
```diff
@@ -0,0 +1,202 @@
+               || 'layout/lightbox' == $this->layout()->getTemplate();
+
+        $getRedirect = function () use ($inLightbox) {
+            // Logged but not authorized (wrong login method etc), close form
```
Onko lomakkeen suora sulkeminen oikea ratkaisu? Pitäisikö näyttää ennemmin ilmoitus siitä, ettei kirjautumistapa kelpaa?

samuli replied:
Käyttöliittymässä ei näytetä lomakkeen avaus-nappia mikäli jokin ehdoista ei toteudu (käyttäjä on kirjautunut Suomi.fillä, ei ole blocklistattu jne). Eli tähän tilanteeseen ei normaalisti pitäisi päätyä. Erikoistilanteita löytyy toki (lomakkeen voi avata manuaalisesti urlin kautta, blocklistaus voi aktivoitua lomakkeen avaamisen jälkeen jne) mutta ajattelin että nämä eivät vaadi tarkempaa hallintaa?

EreMaijala replied:
Jos ei pitäisi tapahtua, niin ok.

EreMaijala commented on module/Finna/src/Finna/Controller/R2ControllerTrait.php:
```diff
@@ -0,0 +1,202 @@
+            }
+        };
+
+        if (!$user = $this->getUser()) {
```
sulut !()

EreMaijala commented on module/Finna/src/Finna/Controller/R2ControllerTrait.php:
```diff
@@ -0,0 +1,202 @@
+     */
+    protected function replaceR2RegisterFormId($formId)
+    {
+        if (! \Finna\Form\Form::isR2RegisterForm($formId, true)) {
```
väli pois ! jälkeen

EreMaijala commented on module/Finna/src/Finna/Controller/R2ControllerTrait.php:
```diff
@@ -0,0 +1,202 @@
+
+        if ($this->formWasSubmitted('submit')) {
+            // Handle submitted registration form
+            $user = $this->getUser();
```
$user on jo haettu aiemmin

EreMaijala commented on module/Finna/src/Finna/Controller/R2ControllerTrait.php:
```diff
@@ -0,0 +1,202 @@
+            $params = $this->params()->fromPost();
+            $form->setData($params);
+
+            if (! $form->isValid()) {
```
väli pois ! jälkeen

EreMaijala commented on module/Finna/src/Finna/Controller/R2ControllerTrait.php:
```diff
@@ -0,0 +1,202 @@
+                );
+            } catch (\Exception $e) {
+                $this->flashMessenger()->addErrorMessage('R2_register_error');
+                $this->flashMessenger()->addErrorMessage($e->getMessage());
```
Yleensä exceptionit logataan, mutta ei näytetä käyttäjälle, ettei vahingossa paljasteta mitään ylimääräistä.

EreMaijala commented on module/Finna/src/Finna/Form/Form.php:
```diff
@@ -497,6 +522,76 @@ public function getFormFields()
     }
 
+    /**
+     * Check if the given form is a R2 registratioin form.
```
typo

EreMaijala replied:
typo: registratioin

EreMaijala commented on module/Finna/src/Finna/Form/Form.php:
```diff
@@ -497,6 +522,76 @@ public function getFormFields()
+     * Check if the given form is a R2 registratioin form.
+     *
+     * @param string  $formId               Form id
+     * @param boolean $checkOnlyNewUserForm Check only new user registration form?
```
Ei tohon kysymysmerkkiä tartte, mutta parempi selitys voisi olla. Olisko siistimpi, jos palauttaisi tiedon siitä, millainen form annettu lomake on booleanin sijaan?

Commit:
Rename blacklist > blocklist

Commit:
Fix typo.

Commit:
Fix typo.

Commit:
Fix typo.

Commit:
Always populate registration form name fields from profile.

Commit:
Log registration exception message. Cleanup.

Commit:
Cleanup.

Commit:
phpcsfixer.

Commit:
Separate R2 registration form handling from Finna forms.

Commit:
Remove unused.

Commit:
Edit comments.

Commit:
Simplify.

Commit:
Cleanup.

Commit:
Edit sample config.

Commit:
Cleanup.

Commit:
Add line break.

Commit:
Fix indentation.

Commit:
phpcsfixer.

Commit:
Update texts.

Commit:
Handle submitted R2 form in R2FeedbackController.

Commit:
Fix form texts.

Commit:
Rename.

Commit:
Upstream changes.

Commit:
Fix long line.

Commit:
Update to work with new REMS api.

EreMaijala commented on module/Finna/src/Finna/Controller/R2FeedbackController.php:
```diff
@@ -0,0 +1,218 @@
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     http://vufind.org   Main Site
+ */
+class R2FeedbackController extends FeedbackController
```
Lisää `implements \Laminas\Log\LoggerAwareInterface` niin ei tartte constructorissa tehdä loggeria.

samuli committed:
Update fi.ini

EreMaijala commented on module/Finna/src/Finna/Controller/R2FeedbackController.php:
```diff
@@ -0,0 +1,218 @@
+{
+    use \VuFind\Log\LoggerAwareTrait;
+
+    protected $formClass = \Finna\Form\R2Form::class;
```
Doc comment puuttuu

EreMaijala commented on module/Finna/src/Finna/Controller/R2FeedbackController.php:
```diff
@@ -0,0 +1,218 @@
+    public function formAction()
+    {
+        $formId = $this->params()->fromRoute('id', $this->params()->fromQuery('id'));
+        if (!\Finna\Form\R2Form::isR2RegisterForm($formId)) {
```
Ota R2Form use-lausekkeella mukaan, niin helpompi lukea ja alempana saa rivityksen nätimmäksi.

EreMaijala commented on module/Finna/src/Finna/Controller/R2FeedbackController.php:
```diff
@@ -0,0 +1,218 @@
+            }
+        }
+
+        $inLightbox
```
Controllereilla on inLightbox()-metodi, käytä sitä.

EreMaijala commented on module/Finna/src/Finna/Controller/R2FeedbackController.php:
```diff
@@ -0,0 +1,218 @@
+            }
+        };
+
+        if (!($user = $this->getUser())) {
```
Voisiko tämä olla jo aiemmin?

EreMaijala commented on module/Finna/src/Finna/Controller/R2FeedbackController.php:
```diff
@@ -0,0 +1,218 @@
+        }
+
+        if ($submitted) {
+            $form = $this->serviceLocator->get(\Finna\Form\R2Form::class);
```
Miksei $this->formClass?

EreMaijala commented on module/VuFind/src/VuFind/Controller/FeedbackController.php:
```diff
@@ -29,6 +29,13 @@
  */
 class FeedbackController extends AbstractBase
 {
+    /**
```
Mergeä uusin dev, näiden pitäisi olla siellä.

Commit:
Merge remote-tracking branch 'origin/dev' into R2-rems-registration

Commit:
Merge branch 'R2-rems-registration' of github.com:samuli/NDL-VuFind2 into R2-rems-registration

Commit:
Code review changes.

- Use LoggerAwareInterface.
- Incorporate upstream changes.
- Exit early when unlogged.
- Use controller method inLIghtbox

Commit:
Remove unused.
